batcher/aligned-sdk/src/communication/messaging.rs

```diff
@@ -55,178 +55,198 @@ pub async fn send_messages(
+        // Save the verification data commitment to read its response later
+        sent_verification_data.push(verification_data);
     }
 
+    info!("All messages sent");
     Ok(sent_verification_data)
 }
 
+// Receives the array of proofs sent
+// Reads the WS responses
+// Matches each response with the corresponding proof sent
+// finishes when the last proof sent receives its response
 pub async fn receive(
     response_stream: Arc<Mutex<ResponseStream>>,
-    ws_write: Arc<Mutex<SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>>>,
-    total_messages: usize,
-    num_responses: Arc<Mutex<usize>>,
-    verification_data_commitments_rev: &mut Vec<VerificationDataCommitment>,
+    mut sent_verification_data: Vec<NoncedVerificationData>,
 ) -> Result<Vec<AlignedVerificationData>, SubmitError> {
     // Responses are filtered to only admit binary or close messages.
     let mut response_stream = response_stream.lock().await;
+    let mut aligned_submitted_data: Vec<AlignedVerificationData> = Vec::new();
+    let last_proof_nonce = get_biggest_nonce(&sent_verification_data);
 
-    let mut aligned_verification_data: Vec<AlignedVerificationData> = Vec::new();
-
+    // read from WS
     while let Some(Ok(msg)) = response_stream.next().await {
+        // unexpected WS close:
         if let Message::Close(close_frame) = msg {
             if let Some(close_msg) = close_frame {
                 return Err(SubmitError::WebSocketClosedUnexpectedlyError(
                     close_msg.to_owned(),
                 ));
             }
             return Err(SubmitError::GenericError(
-                "Connection was closed without close message before receiving all messages"
-                    .to_string(),
+                "Connection was closed before receive() processed all sent messages ".to_string(),
             ));
         }
-        process_batch_inclusion_data(
-            msg,
-            &mut aligned_verification_data,
-            verification_data_commitments_rev,
-            num_responses.clone(),
-        )
-        .await?;
 
-        if *num_responses.lock().await == total_messages {
-            debug!("All messages responded. Closing connection...");
-            ws_write.lock().await.close().await?;
-            return Ok(aligned_verification_data);
+        let batch_inclusion_data_message = handle_batcher_response(msg).await?;
+
+        let related_verification_data = match_batcher_response_with_stored_verification_data(
+            &batch_inclusion_data_message,
+            &mut sent_verification_data,
+        )?;
+
+        let aligned_verification_data =
+            process_batcher_response(&batch_inclusion_data_message, &related_verification_data)?;
+
+        aligned_submitted_data.push(aligned_verification_data);
+        info!("Message response handled succesfully");
+
+        if batch_inclusion_data_message.user_nonce == last_proof_nonce {
```

Comment:
We should quit when we receive the las proof nonce because if that one was included, the previous ones were included as well.
We don't quit after receiving N messages because we can't know how many responses we actually expect. For example, if i send 10 messages with the same nonce, i will receive only 1 response because internally the batcher has interpreted them as replacement messages

Reply:
At the same time, I agree it would be nice to catch any responses if they arrive after the last nonce (which in theory it shouldn't happen)

Reply:
Ohh ok, I thought we were always receiving a response from the batcher, even indicating replacements or errors.
Yes, if this is not a 1-1 relationship, then we can't be sure of how many messages we are expecting.
If we can be 100% sure that once a proof gets included, all proofs with lower nonce must have been included, then I'm ok with the implementation. Anyways, I have a question related with this: [Here](https://github.com/yetanotherco/aligned_layer/blob/1314-fix-sender-receive-batchinclusiondata-or-createnewtaskerror-in-validityresponsemessage/batcher/aligned/src/main.rs#L345) we are submitting multiple proofs and then saving the results. If we stop executing when we find the last proof nonce we will miss the intermediate proof results!

Reply:
The last message the user receives as response is the one of biggest nonce, because they are processed and responded in nonce order, and sent via ws tcp.